src/App.tsx

```diff
@@ -1,24 +1,196 @@
+    }
+  };
+
+  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    addTodo({
+      title: newTodoTitle,
+      completed: false,
+      userId: USER_ID,
+    });
+
+    setNewTodoTitle('');
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const visibleTodos = todos.filter((td) => {
+    switch (filter) {
+      case Filters.Active:
+        return !td.completed;
+      case Filters.Completed:
+        return td.completed;
+      default:
+        return true;
+    }
+  });
+
+  const removeTodo = (todoId: number) => {
+    setProcessedTodos(current => [...current, todoId]);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos((currentTodos) => currentTodos.filter(td => td.id !== todoId));
+      })
+      .catch(() => {
+        setErrorMsg(ErrorMessage.Deleting);
+        setTimeout(() => {
+          setErrorMsg('');
+        }, 3000);
+      })
+      .finally(() => {
+        processedTodos.filter(id => id !== todoId);
+      });
+  };
+
+  const todoUpdate = (todoToUpdate: Todo) => {
+    setProcessedTodos(prev => [...prev, todoToUpdate.id]);
+
+    updateTodo(todoToUpdate)
+      .then(() => {
+        setTodos(
+          todos.map(td => {
+            if (td.id === todoToUpdate.id) {
+              return todoToUpdate;
+            }
+
+            return td;
+          }),
+        );
+      })
+      .catch(() => {
+        setErrorMsg(ErrorMessage.Updating);
+        setTimeout(() => {
+          setErrorMsg('');
+        }, 3000);
+      })
+      .finally(() => {
+        setProcessedTodos(prev => prev.filter(id => id !== todoToUpdate.id));
+      });
+  };
+
+  const completedTodos = todos.filter(td => td.completed);
+  const clearCompleted = () => {
+    completedTodos.forEach(td => {
+      removeTodo(td.id);
+    });
+  };
+
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">React Todo App - Load Todos</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <NewTodoForm
+          todos={visibleTodos}
+          onSubmit={handleSubmit}
+          newTodoTitle={newTodoTitle}
+          setNewTodoTitle={setNewTodoTitle}
+        />
+        {todos.length > 0 && (
+          <FormMain
+            todos={visibleTodos}
+            onRemove={removeTodo}
+            onTodoUpdate={todoUpdate}
+            processedTodos={processedTodos}
+          />
+        )}
+
+        {todos.length > 0 && (
+          <FormFooter
+            todos={todos}
+            filter={filter}
+            setFilter={setFilter}
+            onClearCompleted={clearCompleted}
+          />
+        )}
```

Comment:
if it's the same condition, just combine them
with a fragment i guess
also you can use a shorthand conditional
```suggestion
        {todos.length && (
          <FormMain
            todos={visibleTodos}
            onRemove={removeTodo}
            onTodoUpdate={todoUpdate}
            processedTodos={processedTodos}
          />
          <FormFooter
            todos={todos}
            filter={filter}
            setFilter={setFilter}
            onClearCompleted={clearCompleted}
          />
        )}
```